### PR TITLE
Fix epi weeks

### PR DESCRIPTION
**Pull request by tomeksabala**
Fixed incorrect epi year for epi_configs as dicts (Jordan case).

Refactored epi methods names and extracted epi tests to separate module.
Extended test coverage for epi week computation.

### CODE REVIEW

File: meerkat_abacus/util/__init__.py [hunk 1]
```diff
@@ -50,20 +52,21 @@ def is_child(parent, child, locations):
     return False
 
 
-def epi_week(date):
+def epi_week_for_date(date):
     """
     calculate epi week
```
Comment (Gulfa) [inline meerkat_abacus/util/__init__.py]: Could we include here or something else how we deal with the edge cases. Do we have week 0 or week 53 for example?

File: meerkat_abacus/util/__init__.py [hunk 2]
```diff
@@ -76,21 +79,69 @@ def get_db_engine(db_url=config.DATABASE_URL):
...
+    elif "day" in epi_config:
+        day_of_week = int(epi_config.split(":")[1])
+        first_of_year = datetime(date.year, 1, 1)
+        f_day_of_week = first_of_year.weekday()
+        adjustment = day_of_week - f_day_of_week
+        if adjustment < 0:
+            adjustment = 7 + adjustment
+        return first_of_year + timedelta(days=adjustment)
+    elif isinstance(epi_config, dict):
+        for start_datetime in reversed(sorted(epi_config.values())):
```
Comment (Gulfa) [inline meerkat_abacus/util/__init__.py]: Not sure if this is much better but maybe we can replace reversed(.....) with something like this: 
[ epi_config[year] for year in [year - 1, year, year +1]]
Where year is the year of the current date. We should only have to check these three values I think.

File: meerkat_abacus/util/__init__.py [hunk 3]
```diff
@@ -102,25 +153,53 @@ def epi_week_start_date(year, epi_config=country_config["epi_week"]):
...
+        if epi_config is a dict one can specify custom start dates for epi years
+        e.g.
+        "epi_week": {
+            2011: datetime.datetime(2011, 1, 2),
+            2012: datetime.datetime(2011, 12, 31)
+        }
+    :return: epi year
+    """
+    if isinstance(epi_config, dict):
+        for epi_year, epi_year_start_datetime in reversed(sorted(epi_config.items())):
```
Comment (Gulfa) [inline meerkat_abacus/util/__init__.py]: potentially the same approach here as above

File: meerkat_abacus/test/epi_week_test.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    def test_epi_year_start_for_custom_start_date(self):
+        test_config = {
+            2016: datetime(2016, 1, 2),
+            2017: datetime(2016, 12, 30)
+        }
+        test_data = [
+            {"date": datetime(2016, 3, 5), "expected_year": 2016},
+            {"date": datetime(2016, 12, 31), "expected_year": 2017},
+            {"date": datetime(2017, 4, 24), "expected_year": 2017}
```
Comment (Gulfa) [inline meerkat_abacus/test/epi_week_test.py]: Can we also add a test when date is in the right year, but smaller than start date. So e.g 2016/1/1. As Epi year should then be 2015.

File: meerkat_abacus/test/epi_week_test.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        test_data = [
+            {"date": datetime(2016, 1, 1), "expected_year": 2015},
+            {"date": datetime(2016, 3, 5), "expected_year": 2016},
+            {"date": datetime(2016, 12, 31), "expected_year": 2017},
+            {"date": datetime(2017, 4, 24), "expected_year": 2017}
+        ]
+        for _test in test_data:
+            expected_year = _test["expected_year"]
+            actual_year = epi_year_by_date(_test["date"], epi_config=test_config)
+            self.assertEqual(expected_year, actual_year)
```
Comment (Gulfa) [inline meerkat_abacus/test/epi_week_test.py]: Is the test for the actual epi-week calculation missing? If it is it would be good to include it with all the edge cases with the potential of week 0 and week 53 etc.

Comment (tomeksabala) [reply meerkat_abacus/test/epi_week_test.py]: We have some tests in meerkat_api. I was too lazy to move them to abacus as well. That should have been done. Will update.

File: meerkat_abacus/util/__init__.py (hunk 2 above)
Comment (tomeksabala) [reply meerkat_abacus/util/__init__.py]: I don't think that performance is an issue here. Sorting and reversing a collection of < 50 elements should be very fast. IMO current solution is clearer and simpler.

Comment (Gulfa) [reply meerkat_abacus/util/__init__.py]: Fair enough.
